src/commands.jl

```diff
@@ -177,13 +197,138 @@ function next_line!(@nospecialize(recurse), frame::Frame, istoplevel::Bool=false
+
+    cframe, cpc = ret
+
+where `cframe` is the frame from which execution should continue and `cpc` is the state
+of `cframe` (the program counter, a `BreakpointRef`, or `nothing`).
+"""
+function maybe_reset_frame!(@nospecialize(recurse), frame::Frame, @nospecialize(pc), rootistoplevel::Bool)
+    isa(pc, BreakpointRef) && return leaf(frame), pc
+    if pc === nothing
+        val = get_return(frame)
+        recycle(frame)
+        frame = caller(frame)
+        frame === nothing && return nothing
+        frame.callee = nothing
+        maybe_assign!(frame, val)
+        frame.pc += 1
+        pc = maybe_next_call!(recurse, frame, rootistoplevel && frame.caller===nothing)
+        return maybe_reset_frame!(recurse, frame, pc, rootistoplevel)
+    end
+    return frame, pc
+end
+
+"""
+    ret = debug_command(recurse, frame, cmd, rootistoplevel=false)
+    ret = debug_command(frame, cmd, rootistoplevel=false)
+
+Perform one "debugger" command. `cmd` should be one of:
+
+- "n": advance to the next line
+- "s": step into the next call
+- "c": continue execution until termination or reaching a breakpoint
+- "finish": finish the current frame and return to the parent
+
+or one of the 'advanced' commands
+
+- "nc": step forward to the next call
+- "se": execute a single statement
+- "si": execute a single statement, stepping in if it's a call
+- "sg": step into the generator of a generated function
+
+`rootistoplevel` and `ret` are as described for [`JuliaInterpreter.maybe_reset_frame!`](@ref).
+Unlike other commands, the default setting for `recurse` is `Compiled()`.
+"""
+function debug_command(@nospecialize(recurse), frame::Frame, cmd::AbstractString, rootistoplevel::Bool=false)
+    # ::Union{Val{:nc},Val{:n},Val{:se}},
+    # Need try/catch
+    istoplevel = rootistoplevel && frame.caller === nothing
+    cmd == "nc" && return maybe_reset_frame!(recurse, frame, next_call!(recurse, frame, istoplevel), rootistoplevel)
+    cmd == "n" && return maybe_reset_frame!(recurse, frame, next_line!(recurse, frame, istoplevel), rootistoplevel)
+    if cmd == "si"
+        stmt = pc_expr(frame)
+        cmd = is_call(stmt) ? "s" : "se"
+    end
+    cmd == "se" && return maybe_reset_frame!(recurse, frame, step_expr!(recurse, frame, istoplevel), rootistoplevel)
+    enter_generated = false
+    if cmd == "sg"
+        enter_generated = true
+        cmd = "s"
+    end
+    if cmd == "s"
+        pc = maybe_next_call!(recurse, frame, istoplevel)
+        (isa(pc, BreakpointRef) || pc === nothing) && return maybe_reset_frame!(recurse, frame, pc, rootistoplevel)
+        stmt0 = stmt = pc_expr(frame, pc)
+        if isexpr(stmt, :(=))
+            stmt = stmt.args[2]
+        end
+        ret = evaluate_call!(dummy_breakpoint, frame, stmt; enter_generated=enter_generated)
+        isa(ret, BreakpointRef) && return maybe_reset_frame!(recurse, frame, ret, rootistoplevel)
+        maybe_assign!(frame, stmt0, ret)
+        frame.pc = ret + 1
```

Comment:
There is some issue here when evaluating builtins

```
julia> f(x) = x^2
 f (generic function with 1 method)

julia> frame = JuliaInterpreter.enter_call(f, 3)
Frame for f(x) in Main at REPL[38]:1
  1 1  1 ─ %1 = (Core.apply_type)(Val, 2)
  2 1  │   %2 = (%1)()
  3 1  │   %3 = (Base.literal_pow)(^, x, %2)
⋮
x = 3

julia> frame, bp = JuliaInterpreter.debug_command(JuliaInterpreter.finish_and_return!, frame, "s")
ERROR: MethodError: no method matching +(::Type{Val{2}}, ::Int64)
```

Reply:
Another when the expression is a return statement:

```
> f(x) = x
 f (generic function with 1 method)

> fr = JuliaInterpreter.enter_call(f, 2)
Frame for f(x) in Main at REPL[8]:1
  1 1  1 ─     return x
x = 2

> debug_command(fr, "s")
(frame.framedata).exception_frames = Int64[]
ERROR: MethodError: objects of type Int64 are not callable
Stacktrace:
 [1] #prepare_call#12(::Bool, ::Function, ::Any, ::Array{Any,1}) at C:\Users\Kristoffer\Debugging\JuliaInterpreter\src\construct.jl:216
```

Reply:
fixed